SAMLPortal/Controllers/SetupController.cs

```diff
@@ -0,0 +1,374 @@
+
+			StepFourModel model = new StepFourModel();
+			model.SearchBase = ldapSearchBase != null ? ldapSearchBase : "";
+			model.AdminFilter = ldapAdminFilter != null ? ldapAdminFilter : "uid={0}";
+			model.UserFilter = ldapUserFilter != null ? ldapUserFilter : "uid={0}";
+
+			return View(model);
+		}
+
+		[HttpPost]
+		[Route("4")]
+		public IActionResult StepFour(StepFourModel model)
+		{
+			if (!CheckStep(4))
+			{
+				return Redirect("3");
+			}
+
+			if (ModelState.IsValid)
+			{
+				GlobalSettings.Store("LDAP_SearchBase", model.SearchBase);
+				GlobalSettings.Store("LDAP_AdminFilter", model.AdminFilter);
+				GlobalSettings.Store("LDAP_UsersFilter", model.UserFilter);
+
+				Helpers.ReplaceEnvVariableInFile(GlobalSettings.Get("CONFIG_FILE"), "SP_CONFIG_SETUPASSISTANT_STEP", "5");
+				GlobalSettings.Store("CONFIG_SETUPASSISTANT_STEP", "5");
+			}
+
+			return View(model);
+		}
+
+		[HttpGet]
+		[Route("5")]
+		public IActionResult StepFive()
+		{
+			if (!CheckStep(5))
+			{
+				return Redirect("4");
+			}
+
+			var ldapDisplayName = GlobalSettings.Get("LDAP_Attr_DisplayName");
+			var ldapUID = GlobalSettings.Get("LDAP_Attr_UID");
+			var ldapMemberOf = GlobalSettings.Get("LDAP_Attr_MemberOf");
+			var ldapMail = GlobalSettings.Get("LDAP_Attr_Mail");
+
+			StepFiveModel model = new StepFiveModel();
+			model.DisplayName = ldapDisplayName != null ? ldapDisplayName : "cn";
+			model.UID = ldapUID != null ? ldapUID : "uid";
+			model.MemberOf = ldapMemberOf != null ? ldapMemberOf : "memberOf";
+			model.Mail = ldapMail != null ? ldapMail : "mail";
+
+			return View(model);
+		}
+
+		[HttpPost]
+		[Route("5")]
+		public IActionResult StepFive(StepFiveModel model)
+		{
+			if (!CheckStep(5))
+			{
+				return Redirect("4");
+			}
+
+			if (ModelState.IsValid)
+			{
+				GlobalSettings.Store("LDAP_Attr_DisplayName", model.DisplayName);
+				GlobalSettings.Store("LDAP_Attr_UID", model.UID);
+				GlobalSettings.Store("LDAP_Attr_MemberOf", model.MemberOf);
+				GlobalSettings.Store("LDAP_Attr_Mail", model.Mail);
+
+				Helpers.ReplaceEnvVariableInFile(GlobalSettings.Get("CONFIG_FILE"), "SP_CONFIG_SETUPASSISTANT_STEP", "6");
+				GlobalSettings.Store("CONFIG_SETUPASSISTANT_STEP", "6");
+
+				return Redirect("/");
+			}
+
+			return View(model);
+		}
+
+		/// <summary>
+		/// Returns false if the user purposely miss a step
+		/// </summary>
+		/// <param name="step"></param>
+		/// <returns></returns>
+		private bool CheckStep(int step)
+		{
+			if (step - GlobalSettings.GetInt("CONFIG_SETUPASSISTANT_STEP") <= 1)
+			{
+				return true;
+			}
+			else
+			{
+				return false;
+			}
+		}
+	}
+
+	public class SetupAsyncActionFilter : IAsyncActionFilter
+	{
+		public async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
+		{
+			if (GlobalSettings.GetInt("CONFIG_SETUPASSISTANT_STEP") > 5)
+			{
+				// Setup has been completed and cannot be accessed anymore
+				context.HttpContext.Response.Redirect("/");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to add validation of parameter 'context' before using it.](https://app.codacy.com/manual/zegorax/SAMLPortal/pullRequest?prid=5343548)